python/lsst/sconsUtils/state.py

```diff
@@ -128,6 +130,11 @@ def _initEnvironment():
                 ourEnv[varname] = os.environ[varname]
                 ourEnv[k] = os.environ[k]
 
+    # add <build root>/ups directory to the configuration search path
+    # this allows the .cfg file for the package being built to be found without
+    # requiring <product name>_DIR to be in the env
+    cfgPath.append(os.path.join(SCons.Script.Dir('#').abspath, 'ups'))
```

Comment:
So, I think I've gotten to the bottom of this:
 - SCons *should* be looking only in `_DIR` (and `_DIR_EXTRA`) associated with its dependency tree, but it's not - apparently just because I was lazy when I implemented it.  That should probably be fixed, but I don't think this is currently a big performance hit, simply because it only scales with the number of setup packages, which is not a large number even when a stack contains many versions (since only one can be setup at a time).  If we want to change this, we need to be careful of multi-library packages like boost, which have multiple .cfg files for different libraries within a single EUPS product.
 - There is currently no code to add just the paths for the dependencies needed by a particular package, just the block that adds everything (the one just above the addition this comment is on).  There's nothing in dependencies.py that deals with `_DIR` directories; they're already expected to be in `cfgPath` by then.  The duplication @jhoblitt saw is actually in that very block, due to what looks like a bad merge conflict resolution (note the `append` statement above the `if` block as well as those within it).
 - Unfortunately, I think we might be relying on a byproduct of that accident, because there may be an inconsistency as to whether .cfg files for manually-declared system packages should be in `PRODUCT_DIR_EXTRA/ups` or just `PRODUCT_DIR_EXTRA`.  We probably want to continue to add both of these to `cfgPath`.  But there's obviously no reason to add `PRODUCT_DIR` to `cfgPath` twice.